### PR TITLE
[Push notifications] Add error event

### PR DESCRIPTION
**Pull request by DannyvanderJagt**
> This is PR is an updated version of: https://github.com/facebook/react-native/pull/2336 (closed)

This PR will add an `error` event to `PushNotificationIOS`.

**How does this work?**
The errors from `didFailToRegisterForRemoteNotificationsWithError` will be parsed to the `error` event in javascript. 

**Why is this useful?**
At the moment PushNotificationIOS fails silently and there is no way of telling if something went wrong. 
It would be very helpful to see the errors for debugging and in-app error handling. 

**How to use**
The error event can be used just like the register and the notification events.
The error callback will give the original error object from IOS. 

```js
PushNotificationIOS.addEventListener('error', function(error){
  // Example: {NSLocalizedDescription: "REMOTE_NOTIFICATION_SIMULATOR_NOT_SUPPORTED_NSERROR_DESCRIPTION"}
   console.log('An error occurred: ', error);
});
```

To use the `error` event in javascript add these line to your `AppDelegate.m`:
```objective-c
- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
{
  [RCTPushNotificationManager didFailToRegisterForRemoteNotificationsWithError:error];
}
```

**PR**
This is an new updated version of my 7 month closed old PR: https://github.com/facebook/react-native/pull/2336. I couldn't get the old branch updated without destroying the PR. I forked the master of react-native today and added back and updated all the code of the old PR. 

**TODO: Docs**
This event should be added to the docs and the website. Can someone tell me how I can accomplish this?

**TODO: Review**
Objective-C is not a language I use often. It would be appreciated if someone could review this PR :)

### CODE REVIEW

**Comment by DannyvanderJagt** [comment]
@satya164 Instead of re-oping the old PR (https://github.com/facebook/react-native/pull/2336) i created a new PR. I couldn't get the old branch updated without destroying the PR. I forked the master of react-native today and added back and updated all the code of the old PR.

**Comment by facebook-github-bot-6** [comment]
By analyzing the blame information on this pull request, we identified @slycoder and @nicklockwood to be potential reviewers.

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 1]
```diff
@@ -153,6 +167,12 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
                                               body:notification.userInfo];
 }
 
+- (void)handleRemoteNotificationRegisteredError:(NSNotification *)notification
+{
+  [_bridge.eventDispatcher sendDeviceEventWithName:@"remoteNotificationsRegisteredError"
+                                              body:[notification userInfo]];
```
Comment (alloy) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Judging by the preceding notification handlers, dot-notation is preferred: `notification.userInfo`.

File: Libraries/PushNotificationIOS/PushNotificationIOS.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ var _initialNotification = RCTPushNotificationManager &&
 
 var DEVICE_NOTIF_EVENT = 'remoteNotificationReceived';
 var NOTIF_REGISTER_EVENT = 'remoteNotificationsRegistered';
+var NOTIF_REGISTER_ERROR_EVENT = 'remoteNotificationsRegisteredError';
```
Comment (alloy) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: This name doesn’t really read nicely. What about a name like `remoteNotificationsRegistrationError` (and in all other places in the code) ?

File: Libraries/PushNotificationIOS/PushNotificationIOS.js [hunk 3]
```diff
@@ -178,6 +179,13 @@ class PushNotificationIOS {
           handler(registrationInfo.deviceToken);
         }
       );
+    } else if(type === 'error'){
```
Comment (alloy) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: Add a space between `if(`.

**Comment by alloy** [comment]
@DannyvanderJagt Did a bit of review, otherwise it looks good to me. Hope it helps.

**Comment by DannyvanderJagt** [comment]
@alloy Thank you! I will update the PR in a few minutes.

**Comment by alloy** [comment]
@nicklockwood Code-wise this seems good to me :+1:

**Comment by facebook-github-bot-2** [comment]
@DannyvanderJagt updated the pull request.

**Comment by JAStanton** [comment]
Hey, just noticed this still hasn't landed yet. Is there anything that I can do to help push this PR along?

**Comment by DannyvanderJagt** [comment]
We are waiting for a review from a contributor I believe. I think that there are busy :)

**Comment by nicklockwood** [comment]
I think this will need some rebasing due to the event emitter changes, but I'll see what I can do.

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

**Comment by DannyvanderJagt** [comment]
@nicklockwood Thank you! Let me know if there is anything i can do.

File: Libraries/PushNotificationIOS/RCTPushNotificationManager.m [hunk 4]
```diff
@@ -135,6 +140,15 @@ + (void)didReceiveLocalNotification:(UILocalNotification *)notification
                                                     userInfo:details];
 }
 
+
+
++ (void)didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
```
Comment (javache) [inline Libraries/PushNotificationIOS/RCTPushNotificationManager.m]: Can you the update the documentation in https://github.com/facebook/react-native/blob/c6020a0ef4880bead15a0ed2f4058a70cc9b6ad5/Libraries/PushNotificationIOS/PushNotificationIOS.js#L73 to mention this method.

**Comment by facebook-github-bot-7** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @nicklockwood as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by nevir** [comment]
@DannyvanderJagt looks like this could use a rebase off of master; it has some conflicts w/ the new event emitter code

**Comment by DannyvanderJagt** [comment]
@nevir @nicklockwood I will rebase this PR in the next few days.

**Comment by denisw** [comment]
I see this pull request was closed. Is there a follow-up PR? If yes, it would be good to link to it from here.

**Comment by nevir** [comment]
Just pushed an updated version of it via https://github.com/facebook/react-native/pull/9650

**Comment by DannyvanderJagt** [comment]
@nevir Thank you for picking up this PR!

**Pull request closed without merging**